src/Workspaces/Core/Portable/Workspace/Solution/ProjectIdMapping.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Frozen;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis;
+
+/// <summary>
+/// Mapping from ProjectId to values of type T. Set of possible ProjectID keys are determined at construction.
+/// Population of values are done through calls to GetOrAdd.
+/// </summary>
+/// <typeparam name="T"></typeparam>
+internal struct ProjectIdMapping<T>
```

Comment:
seems like this can be a readonly struct (just don't assign it into a readonly-field :)).

Reply:
Moved away from struct as there is deferred initialization going on now